### PR TITLE
feat: Add ListUsers method to retrieve usernames for a service (#133)

### PR DESCRIPTION
**Pull request by ayoub3bidi**
Implements a new `ListUsers(service string) ([]string, error)` method that retrieves all usernames associated with a given service **without exposing the secret values**. This addresses the feature request to allow CLI tools and applications to discover which credentials are stored for a service.

## Motivation

As described in #133, when using `go-keyring` in CLI tools that manage multiple third-party tokens (e.g., GitHub, AWS, etc.), users need a way to see which secrets the tool is managing without manually checking their system keychain. This feature enables better UX by allowing applications to list available credentials.

**Example use case**: A CLI tool with service name `my-cli` that manages tokens for multiple providers (GitHub, GitLab, AWS). The tool can now call `ListUsers("my-cli")` to show users which providers they have configured.

### Core Interface
- **[keyring.go](keyring.go#L30)**: Added `ListUsers(service string) ([]string, error)` to `Keyring` interface
- **[keyring.go](keyring.go#L53-L56)**: Added package-level `ListUsers` function

#### macOS ([keyring_darwin.go](keyring_darwin.go#L138-L187))
- Parses `security dump-keyring` output to extract account names
- Filters by service name and deduplicates results
- Returns empty slice for non-existent services

#### Linux/Unix ([keyring_unix.go](keyring_unix.go#L180-L221))
- Leverages existing `findServiceItems` helper to query D-Bus Secret Service
- Retrieves username attributes from items
- Added helper method `GetItemAttributes` to [secret_service/secret_service.go](secret_service/secret_service.go#L243-L257)
- Gracefully handles missing services (returns empty slice)

#### Windows ([keyring_windows.go](keyring_windows.go#L96-L123))
- Uses `wincred.List()` to enumerate credentials
- Filters by service prefix and extracts usernames
- Consistent with `DeleteAll` implementation pattern

#### Mock & Fallback
- **[keyring_mock.go](keyring_mock.go#L62-L73)**: Iterates over in-memory store
- **[keyring_fallback.go](keyring_fallback.go#L29-L31)**: Returns `ErrUnsupportedPlatform`

### Tests

Added four comprehensive test cases to [keyring_test.go](keyring_test.go#L186-L318):

- `TestListUsers`: Multiple users per service
- `TestListUsersEmpty`: Non-existent service returns empty slice
- `TestListUsersSingleUser`: Single user edge case
- `TestListUsersMultipleServices`: Service isolation verification

## Test Results

```
$ go test -v
=== RUN   TestListUsers
--- PASS: TestListUsers (0.05s)
=== RUN   TestListUsersEmpty
--- PASS: TestListUsersEmpty (0.00s)
=== RUN   TestListUsersSingleUser
--- PASS: TestListUsersSingleUser (0.02s)
=== RUN   TestListUsersMultipleServices
--- PASS: TestListUsersMultipleServices (0.05s)
PASS
ok      github.com/zalando/go-keyring   0.270s
```

- All 23 tests passing (4 new + 19 existing)  
- No regressions  
- Build successful

## Usage Example

```go
package main

import (
    "fmt"
    "log"
    
    "github.com/zalando/go-keyring"
)

func main() {
    service := "my-cli"
    
    // Store some credentials
    keyring.Set(service, "github", "ghp_token123")
    keyring.Set(service, "gitlab", "glpat_token456")
    keyring.Set(service, "aws", "aws_secret789")
    
    // List all configured providers
    users, err := keyring.ListUsers(service)
    if err != nil {
        log.Fatal(err)
    }
    
    fmt.Println("Configured providers:")
    for _, user := range users {
        fmt.Printf("  - %s\n", user)
    }
    // Output:
    // Configured providers:
    //   - github
    //   - gitlab
    //   - aws
}
```

## Backward Compatibility

**No breaking changes**
- Purely additive API change
- All existing code continues to work without modification
- Optional feature that doesn't affect current users

### Naming Decision
As suggested by @szuecs in #133, the method is named `ListUsers` instead of `List` for clarity and to avoid ambiguity about what is being listed.

### Platform-Specific Behavior
- **Empty service**: All implementations return empty slice (not error)
- **Non-existent service**: Returns empty slice consistently across platforms
- **Order**: Usernames returned in arbitrary order (platform-dependent)
- **Deduplication**: All implementations deduplicate usernames using maps

### Security
- Returns only usernames, never secret values
- Follows existing platform security patterns
- Respects keychain/credential manager locking mechanisms

## Checklist

- [x] Implementation complete for all platforms
- [x] Tests added and passing
- [x] No regressions in existing tests
- [x] Build successful
- [x] Documentation updated
- [x] Backward compatible

### CODE REVIEW

**Comment by szuecs** [comment]
@ayoub3bidi thanks for the PR for me looks like a nice addition.
I would like that the Example you specified here could be an Example that is tested in Go. The Example will be rendered in go docs and it runs by the test suite to make sure the code produces the output. I don't know if it's easily possible but I guess you can a mock implementation.
Last @AlexanderYastrebov has a good review. Let's try to fix the findings.

File: keyring_darwin.go [hunk 1]
```diff
@@ -135,6 +135,79 @@ func (k macOSXKeychain) DeleteAll(service string) error {
...
+		return nil, err
+	}
+
+	// Parse dump-keychain output. Format:
+	// keychain: "/Users/username/Library/Keychains/login.keychain-db"
+	//     class: "genp"
+	//     attributes:
+	//         "svce"<blob>="service-name"
+	//         "acct"<blob>="account-name"
+	valueOf := func(s, prefix string) string {
```
Comment (szuecs) [inline keyring_darwin.go]: We define a function here that is used only for "keychain" line parsing -> do not define a function for this.

File: keyring_darwin.go [hunk 2]
```diff
@@ -135,6 +135,79 @@ func (k macOSXKeychain) DeleteAll(service string) error {
...
+			currentSvc = ""
+			currentAcct = ""
+		case strings.Contains(line, `"svce"`):
+			if idx := strings.Index(line, `="`); idx != -1 {
+				start := idx + 2
+				if end := strings.Index(line[start:], `"`); end != -1 {
+					currentSvc = line[start : start+end]
+				}
+			}
+		case strings.Contains(line, `"acct"`):
```
Comment (szuecs) [inline keyring_darwin.go:176]: the body of this switch is the same as the one above. Here it would make sense to define a function.

File: keyring_unix.go [hunk 3]
```diff
@@ -177,6 +177,52 @@ func (s secretServiceProvider) DeleteAll(service string) error {
...
+
+	svc, err := ss.NewSecretService()
+	if err != nil {
+		return nil, err
+	}
+
+	// Find all items for the service
+	items, err := s.findServiceItems(svc, service)
+	if err != nil {
+		// If service has no items, return empty list instead of error
```
Comment (szuecs) [inline keyring_unix.go]: this we can see in code, the question is why do we do this

File: keyring_unix.go [hunk 4]
```diff
@@ -177,6 +177,52 @@ func (s secretServiceProvider) DeleteAll(service string) error {
...
+	if service == "" {
+		return []string{}, nil
+	}
+
+	svc, err := ss.NewSecretService()
+	if err != nil {
+		return nil, err
+	}
+
+	// Find all items for the service
```
Comment (szuecs) [inline keyring_unix.go]: This we can see, drop the comment please. No obvious comments in code, thanks. Please cleanup everywhere in your change proposal.

File: keyring_unix.go [hunk 5]
```diff
@@ -177,6 +177,52 @@ func (s secretServiceProvider) DeleteAll(service string) error {
...
+	items, err := s.findServiceItems(svc, service)
+	if err != nil {
+		// If service has no items, return empty list instead of error
+		if err == ErrNotFound {
+			return []string{}, nil
+		}
+		return nil, err
+	}
+
+	// Extract usernames from items
```
Comment (szuecs) [inline keyring_unix.go]: drop this

File: keyring_unix.go [hunk 6]
```diff
@@ -177,6 +177,52 @@ func (s secretServiceProvider) DeleteAll(service string) error {
...
+	var users []string
+	seenUsers := make(map[string]bool)
+
+	for _, item := range items {
+		// Get item attributes
+		attrs, err := svc.GetItemAttributes(item)
+		if err != nil {
+			continue // Skip items we can't read
```
Comment (szuecs) [inline keyring_unix.go]: the comment tells what "continue" does, but we are interested in the why

File: keyring_unix.go [hunk 7]
```diff
@@ -177,6 +177,52 @@ func (s secretServiceProvider) DeleteAll(service string) error {
...
+		}
+		return nil, err
+	}
+
+	// Extract usernames from items
+	var users []string
+	seenUsers := make(map[string]bool)
+
+	for _, item := range items {
+		// Get item attributes
```
Comment (szuecs) [inline keyring_unix.go]: What else would `GetItemAttributes` return?

File: keyring_unix.go [hunk 8]
```diff
@@ -177,6 +177,52 @@ func (s secretServiceProvider) DeleteAll(service string) error {
...
+		attrs, err := svc.GetItemAttributes(item)
+		if err != nil {
+			continue // Skip items we can't read
+		}
+
+		// Extract username from attributes
+		if username, ok := attrs["username"]; ok {
+			if !seenUsers[username] {
+				seenUsers[username] = true
+				users = append(users, username)
```
Comment (szuecs) [inline keyring_unix.go]: please drop seenUsers and use https://pkg.go.dev/slices#Compact to get the unique list of users.

File: keyring_unix.go [hunk 9]
```diff
@@ -177,6 +177,52 @@ func (s secretServiceProvider) DeleteAll(service string) error {
...
+	seenUsers := make(map[string]bool)
+
+	for _, item := range items {
+		// Get item attributes
+		attrs, err := svc.GetItemAttributes(item)
+		if err != nil {
+			continue // Skip items we can't read
+		}
+
+		// Extract username from attributes
```
Comment (szuecs) [inline keyring_unix.go]: drop the comment

**Comment by szuecs** [comment]
@ayoub3bidi please drop all non-godoc comments and all comments that are not telling WHY, we do something.
Everywhere: in tests and in production code.

There are some comments to have better code quality.

**Comment by ayoub3bidi** [comment]
@szuecs I have refactored macOS parsing, replaced seenUsers with slices. Compact on Linux/Windows, removed non-WHY comments from production code and ListUsers tests.

**Review by szuecs — APPROVED** [review]

**Comment by neilotoole** [comment]
@ayoub3bidi @szuecs: if it's any motivation (helping solve a long-standing [sq](https://github.com/neilotoole/sq) feature-gap that's currently being worked on), we would love to see this PR land. https://github.com/neilotoole/sq/issues/715

**Comment by szuecs** [comment]
> @ayoub3bidi @szuecs: if it's any motivation (helping solve a long-standing [sq](https://github.com/neilotoole/sq) feature-gap that's currently being worked on), we would love to see this PR land. https://github.com/neilotoole/sq/issues/715

The second maintainer will be back next week.
I already approved and @mikkeloscar will check as soon as he's back and has time to review this pr.
So no worries we will review every pr that makes sense for the project and most of them are merged.

**Comment by mikkeloscar** [comment]
:+1:

**Comment by neilotoole** [comment]
@szuecs @ayoub3bidi @mikkeloscar: I'm working off your master branch, and can confirm, looks good so far. Much obliged.

**Comment by szuecs** [comment]
Thanks @neilotoole for reporting back!
